builtin/mainmenu/settings/components.lua

```diff
@@ -0,0 +1,378 @@
+			local fs = "label[0,0.1;" .. get_label(setting) .. "]"
+
+			fs = fs .. ("dropdown[0,0.3;%f,0.8;%s;%s;%d]"):format(
+				avail_w, setting.name, table.concat(items, ","), selected_idx, value)
+
+			return fs, 1.1
+		end,
+
+		on_submit = function(self, fields)
+			local old_value = core.settings:get(setting.name) or setting.default
+			local value = fields[setting.name]
+			if value == nil or value == old_value then
+				return false
+			end
+
+			core.settings:set(setting.name, value)
+			return true
+		end,
+	}
+end
+
+
+function make.path(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = core.settings:get(setting.name) or setting.default
+			self.changed = value ~= setting.default
+
+			local fs = ("field[0,0.3;%f,0.8;%s;%s;%s]"):format(
+				avail_w - 3, setting.name, get_label(setting), value)
+			fs = fs .. ("button[%f,0.3;1.5,0.8;%s;%s]"):format(avail_w - 3, "pick_" .. setting.name, fgettext("Browse"))
+			fs = fs .. ("button[%f,0.3;1.5,0.8;%s;%s]"):format(avail_w - 1.5, "set_" .. setting.name, fgettext("Set"))
+
+			return fs, 1.1
+		end,
+
+		on_submit = function(self, fields)
+			local dialog_name = "dlg_path_" .. setting.name
+			if fields["pick_" .. setting.name] then
+				local is_file = setting.type ~= "path"
+				core.show_path_select_dialog(dialog_name,
+					is_file and fgettext_ne("Select file") or fgettext_ne("Select directory"), is_file)
+				return true
+			end
+			if fields[dialog_name .. "_accepted"] then
+				local value = fields[dialog_name .. "_accepted"]
+				if value ~= nil then
+					core.settings:set(setting.name, value)
+				end
+				return true
+			end
+			if fields["set_" .. setting.name] then
+				local value = fields[setting.name]
+				if value ~= nil then
+					core.settings:set(setting.name, value)
+				end
+				return true
+			end
+		end,
+	}
+end
+
+
+function make.v3f(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = vector.from_string(core.settings:get(setting.name) or setting.default)
+			self.changed = value ~= vector.from_string(setting.default)
+
+			-- Allocate space for "Set" button
+			avail_w = avail_w - 1
+
+			local fs = "label[0,0.1;" .. get_label(setting) .. "]"
+
+			local field_width = (avail_w - 3*0.25) / 3
+
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				0, field_width, setting.name .. "_x", "X", value.x)
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				field_width + 0.25, field_width, setting.name .. "_y", "Y", value.y)
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				2 * (field_width + 0.25), field_width, setting.name .. "_z", "Z", value.z)
+
+			fs = fs .. ("button[%f,0.6;1,0.8;%s;%s]"):format(avail_w, "set_" .. setting.name, fgettext("Set"))
+
+			return fs, 1.4
+		end,
+
+		on_submit = function(self, fields)
+			if fields["set_" .. setting.name] then
```

Comment:
```suggestion
			if fields["set_" .. setting.name] or
					fields.key_enter_field == setting.name .. "_x" or
					fields.key_enter_field == setting.name .. "_y" or
					fields.key_enter_field == setting.name .. "_z" then
```
Again, if we want to allow Enter key.